### PR TITLE
Re-correct TSAN integration

### PR DESCRIPTION
**Pull request by Keno**
Looks like #36929 got reverted. Not sure what happened there. Re-apply it to fix TSAN.

### CODE REVIEW

**Comment by Keno** [comment]
Will fix it in the morning.

**Comment by vtjnash** [comment]
In certain modes, we copy this state and expect to have certain properties. Need to look closer to see if if this is okay with all the various option combinations.

File: src/task.c [hunk 1]
```diff
@@ -357,9 +372,9 @@ static void ctx_switch(jl_task_t *lastt)
                 t->sticky = 1;
                 t->bufsz = 0;
                 if (always_copy_stacks)
-                    memcpy(&t->copy_stack_ctx, &ptls->copy_stack_ctx, sizeof(t->copy_stack_ctx));
+                    memcpy(&t->copy_stack_ctx, &ptls->copy_stack_ctx, sizeof(t->copy_stack_ctx) - sizeof(t->copy_stack_ctx.tsan_state));
                 else
-                    memcpy(&t->ctx, &ptls->base_ctx, sizeof(t->ctx));
+                    memcpy(&t->ctx, &ptls->base_ctx, sizeof(t->ctx) - sizeof(t->ctx.tsan_state));
```
Comment (vtjnash) [inline src/task.c:377]:
```suggestion
                    memcpy(&t->ctx, &ptls->base_ctx, sizeof(t->ctx.uc_mcontext));
```

File: src/task.c [hunk 2]
```diff
@@ -357,9 +372,9 @@ static void ctx_switch(jl_task_t *lastt)
                 t->sticky = 1;
                 t->bufsz = 0;
                 if (always_copy_stacks)
-                    memcpy(&t->copy_stack_ctx, &ptls->copy_stack_ctx, sizeof(t->copy_stack_ctx));
+                    memcpy(&t->copy_stack_ctx, &ptls->copy_stack_ctx, sizeof(t->copy_stack_ctx) - sizeof(t->copy_stack_ctx.tsan_state));
```
Comment (vtjnash) [inline src/task.c:375]:
```suggestion
                    memcpy(&t->copy_stack_ctx.uc_mcontext, &ptls->copy_stack_ctx.uc_mcontext, sizeof(t->copy_stack_ctx.uc_mcontext));
```

But I'm not sure if `uc_mcontext` is something C can deal with, or if it needed to be wrapped in a struct to be something that C could create access.

Comment (vtjnash) [reply src/task.c:375]: But note that this is in the `COPY_STACKS` code path, so currently we disable this code path entirely if TSAN is enabled

**Comment by vtjnash** [comment]
I will fix and update this, with other fixes

**Pull request closed without merging**